### PR TITLE
[WIP] export ONNX model by components

### LINKED ISSUE #241: Export ONNX fixed input size and export of backbone seperate
Hi,

On certain platforms it is not possible to run the ONNX model "as is" due to version issues and that dynamic tensors are not often supported in older versions of ONNX software.

The work around to these issues are:

a) Set the input size (w,h) used in the export of the ONNX model to a fixed size via a command line option
b) Set option to export all or parts of the network:
1) Entire network
2) Backbone (ShuffleNet/ResNet50) only
3) Heads CIF/CAF only

Thanks.

### PR DESCRIPTION
**Pull request by svenkreiss**
Fixes #241

### CODE REVIEW

File: openpifpaf/export_onnx.py [hunk 1]
```diff
@@ -87,6 +87,41 @@ def apply(model, outfile, verbose=True):
...
+        outfile + '.base.onnx', verbose=verbose,
+        input_names=['input_batch'], output_names=['features'],
+        keep_initializers_as_inputs=True,
+        # opset_version=10,
+        do_constant_folding=True,
+        export_params=True,
+    )
+    output_files.append(outfile + '.base.onnx')
+
+    for head_i, head in enumerate(model.head_nets):
```
Comment (mdenna-nviso) [inline openpifpaf/export_onnx.py:104]: Could we put all the heads in a single exported networks?
Having a separate net for each head complicates a lot postprocessing as we currently  have no way to have one postprocessing code taking output from two networks.

**Comment by mdenna-nviso** [comment]
I tried the export by components.
I get one base network and three head networks. Not clear to me why we get 3 heads, none of which contains **cif** or **caf**:
image

**Comment by svenkreiss** [comment]
Head0 is cif, Head1 is caf. Ignore Head2.
I cannot read what is in the image.

File: openpifpaf/export_onnx.py (hunk 1 above)
Comment (svenkreiss) [reply openpifpaf/export_onnx.py:104]: But that is what we had before no? Without the breakdown by components? Maybe I am still misunderstanding what you need.

**Comment by mdenna-nviso** [comment]
I updated the post with a higher resolution image.
The issue is that it doesn't end with the CAF tensor as I expected.. Am I missing something?

**Comment by svenkreiss** [comment]
Indeed, it misses a post processing step. Thanks.
Will look into it.

File: openpifpaf/export_onnx.py (hunk 1 above)
Comment (mdenna-nviso) [reply openpifpaf/export_onnx.py:104]: Ah sorry, I try to explain better.
The ideal would be to split the network in two parts:
- base (contains the backbone)
- heads (contains all the heads)

Having a separate network for each head could in principle allow to parallelize the execution, but they are so small the the benefit is negligible, so I think it is better to keep all the heads together as this allows to avoid changes in the structure of tje postprocessing code.

Comment (svenkreiss) [reply openpifpaf/export_onnx.py:104]: Why split the base and the head then?

Comment (mdenna-nviso) [reply openpifpaf/export_onnx.py:104]: The base is the most compute-intensive part of the network. This is the part that is definitely worth running on accelerated HW.
The heads are quite small but contains "tricky" operators that are often not supported well or not supported at all by HW accelerators. Splitting the network in two allows to run the backbone on accelerated HW, while the heads part can be implemented in SW so that we have more flexibility in supporting the required node types.

Comment (svenkreiss) [reply openpifpaf/export_onnx.py:104]: Before we had an in-place add operation that created an ATen op that wasn't supported but that is fixed. Otherwise the head should be the simplest part of the network: a convolution, some slicing and some constant offsets to add. Do you have a specific op that is still "tricky"?

Comment (mdenna-nviso) [reply openpifpaf/export_onnx.py:104]: I will check better tomorrow but if I remember correctly Gather node for example is not supported well in tensor-rt

Comment (svenkreiss) [reply openpifpaf/export_onnx.py:104]: Interesting. Didn't know that. Thanks. Have a good evening.

Comment (svenkreiss) [reply openpifpaf/export_onnx.py:104]: Any new insight into which operator is the problematic one ... or two?

Comment (mdenna-nviso) [reply openpifpaf/export_onnx.py:104]: Hi Sven thanks, we are double-checking this,  I will update here when we have more info.

Comment (mdenna-nviso) [reply openpifpaf/export_onnx.py:104]: Looks like Slice layer with negative index is not supported:

```
(Unnamed Layer* 175) [Slice]: slice is out of input range
While parsing node number 178 [Slice]:
ERROR: onnx2trt_utils.hpp:347 In function convert_axis:
[8] Assertion failed: axis >= 0 && axis < nbDims
[2020-07-16 12:07:42.362] [error] error: could not parse onnx engine
[2020-07-16 12:07:42.363] [error] error: could not deserialize or build engine

```
In effect "ends" is -1. This seems a simple change, if this is something you can do I can retry quickly.

Comment (svenkreiss) [reply openpifpaf/export_onnx.py:104]: This seems to be messed up in ONNX TensorRT.

Regardless, PyTorch also tried to be smart about intermediate shapes. Although the upsampled feature map sizes are always static, it tried to obtain them from the graph leading to lots of extra nodes in ONNX. I just pushed a commit that removes a lot of nodes from the head networks in ONNX. The heads are fairly simple now. The shapes should all be static positive numbers now. Let me know how that goes.

image

Comment (mdenna-nviso) [reply openpifpaf/export_onnx.py:104]: Thanks, we will give this a try and let you know!

Comment (svenkreiss) [reply openpifpaf/export_onnx.py:104]: It works now?

Comment (mdenna-nviso) [reply openpifpaf/export_onnx.py:104]: Hi Sven, yes model import works correctly now in tensor-rt. Unfortunately inference execution fails, we are now checking if this is an issue of the model itself or there is some issue in the way we use tensor-rt. I'll update you ASAP
Thanks!

Comment (mdenna-nviso) [reply openpifpaf/export_onnx.py:104]: So.. model import works correctly in tensorrt, but inference fails. We get this error message:

```
NVMEDIA_DLA :  308, ERROR: setInputTensorDesc failed 
NVMEDIA_DLA :  446, ERROR: SetInputTensorDesc failed for tensor: 7. status: 0x0.
NVMEDIA_DLA : 1914, ERROR: BindInputTensorArgs failed (Input). status: 0x7.
../rtExt/dla/native/dlaUtils.cpp (212) - DLA Error in submit: 7 (Failure to submit program to DLA engine.)
FAILED_EXECUTION: std::exception

```
If we keep our code the same but use just the "head" part of the network, then the inference also completes successfully (I still have to analye the resulting tensor to be sure it contains what we expect).

We are still trying to understand what is going wrong, but from what we see a split of the network in two parts (base + head as discussed) would allow us to run the (heavy) base in tensorrt and the rest in LPDNN right now.

For your reference i attach the log result in both cases.

Comment (mdenna-nviso) [reply openpifpaf/export_onnx.py:104]: For some reason file upload fails.
I copy here the success case (running the base only):

```
Detected 1 inputs and 1 output network tensors.
[2020-07-21 17:14:34.121] [info] serialize engine to 
[2020-07-21 17:14:34.121] [warning] empty engine file name, skip save
[2020-07-21 17:14:34.125] [info] create execute context and malloc device memory...
[2020-07-21 17:14:34.125] [info] init engine...
[2020-07-21 17:14:34.137] [info] malloc device memory
nbBingdings: 2
[2020-07-21 17:14:34.138] [info] input: 
[2020-07-21 17:14:34.138] [info] binding bindIndex: 0, name: input_batch, size in byte: 150156
[2020-07-21 17:14:34.138] [info] binding dims with 3 dimemsion
3 x 97 x 129   
[2020-07-21 17:14:34.138] [info] output: 
[2020-07-21 17:14:34.138] [info] binding bindIndex: 1, name: features, size in byte: 516096
[2020-07-21 17:14:34.138] [info] binding dims with 3 dimemsion
2048 x 7 x 9   
Binding dims (size) for input_batch ind 0 is {3, 97, 129} (150156)
Binding dims (size) for features ind 1 is {2048, 7, 9} (516096)
[2020-07-21 17:14:34.150] [info] net forward takes 9.46506 ms

```

Comment (mdenna-nviso) [reply openpifpaf/export_onnx.py:104]: And here the failure case (running the entire network):

```
Detected 1 inputs and 1 output network tensors.
[2020-07-21 17:14:34.121] [info] serialize engine to 
[2020-07-21 17:14:34.121] [warning] empty engine file name, skip save
[2020-07-21 17:14:34.125] [info] create execute context and malloc device memory...
[2020-07-21 17:14:34.125] [info] init engine...
[2020-07-21 17:14:34.137] [info] malloc device memory
nbBingdings: 2
[2020-07-21 17:14:34.138] [info] input: 
[2020-07-21 17:14:34.138] [info] binding bindIndex: 0, name: input_batch, size in byte: 150156
[2020-07-21 17:14:34.138] [info] binding dims with 3 dimemsion
3 x 97 x 129   
[2020-07-21 17:14:34.138] [info] output: 
[2020-07-21 17:14:34.138] [info] binding bindIndex: 1, name: features, size in byte: 516096
[2020-07-21 17:14:34.138] [info] binding dims with 3 dimemsion
2048 x 7 x 9   
Binding dims (size) for input_batch ind 0 is {3, 97, 129} (150156)
Binding dims (size) for features ind 1 is {2048, 7, 9} (516096)
[2020-07-21 17:14:34.150] [info] net forward takes 9.46506 ms

```

Comment (svenkreiss) [reply openpifpaf/export_onnx.py:104]: Failure and success seem to have identical output. Was that a copy&paste fail?

**Comment by svenkreiss** [comment]
I have just grouped all the heads together in the output and fixed the postprocessing step in the export. It only occured to me now, but instead of separating the heads, the export that is called "heads" is a full model export where the base network is the identity.

Running the export with `--model-components` will give you a `.base.onnx` and `.heads.onnx` file. Please let me know how that goes.

**Comment by svenkreiss** [comment]
Did this work?

**Comment by svenkreiss** [comment]
Hello @mdenna-nviso , I looked into a prototype for coreml and found that strided slices are problematic for iOS13. The coreml export actually works, but the model produces wrong results. That's supposedly fixed in iOS14 and the latest coremltools beta. The lacking support in iOS13 is problematic both for the shufflenet backbone and the head networks. 

I am working on a major rewrite of OpenPifPaf for v0.12 that will make strided slices optional and have a unit test for coreml export. That should simplify this process.

**Comment by miljanvs** [comment]
> Did this work? @mdenna-nviso

Hi @svenkreiss , we haven't got it working yet. The split of the networks by @mdenna-nviso seemed successful but now we are struggling with the post-processing. Since the post-processing is quite complex it is difficult to debug. And we are not sure whether we are referring to the right post-processing python code.

We would like to set-up a python golden-reference for our debugging but we are not sure how to find the appropriate pytorch model. The ONNX model we split and we are using in our experiments has the following structure (at the bottom):

Screenshot 2020-09-11 at 12 36 21

**Comment by svenkreiss** [comment]
Hello @miljanvs , can you run `openpifpaf.predict` with your model before exporting it? Can you extract your reference from there? Which base network are you using? I can try to make you a pretrained model for the dev branch. I have a working coreml model there.

**Comment by miljanvs** [comment]
> Hello @miljanvs , can you run `openpifpaf.predict` with your model before exporting it? Can you extract your reference from there? Which base network are you using? I can try to make you a pretrained model for the dev branch. I have a working coreml model there.

Hi @svenkreiss , thanks for your answer. We have been using openpifpaf.predict but as I mentioned we have lost track of the pytorch version of the model--the one we split into base/heads (for the reason of TensorRt that @mdenna-nviso  previously explained) is different from the one we use with openpifpaf.predict (as you can see the heads part above). The base is still resnet50-011 but the versions seem not to match and possibly that is a reason why the post-processing is not working. We are not using coreml in this case but TensorRT for the base part and our internal engine for the heads part. If you could help to align the models and establish a golden python reference, it would be great. Thanks.

**Comment by svenkreiss** [comment]
If you feel adventurous, since this morning, there is a test for this in the dev branch:
https://github.com/vita-epfl/openpifpaf/blob/dev/tests/test_onnx_export.py#L34
The dev branch contains major changes to the layout of the head networks that should simplify ONNX graphs.
Everything is preliminary and subject to change in dev, but it has pretrained models for shufflenetv2k16w (not final) and resnet50.

The test does a full loop: export an onnx model, create predictions with pytorch (original model) and onnxruntime (onnx model), compare outputs.

Even if you don't want to switch to dev, this test case might still be helpful.

**Comment by miljanvs** [comment]
Thanks @svenkreiss , it is a great thing that now the onnx export is verified through the test loop, we will try it. We felt slightly adventurous and tried with the latest network from the dev branch. However, we were not able to import the exported onnx file because the heads part has changed with respect to the architecture shown in the previous image in my original message above (the last one that was digestible for our engine):

Inputs = ['input_batch'], Outputs = ['cif', 'caf', '564']
Import error: Unsupported permutation:[0, 2, 1, 3, 4]: 506
Import error: Unsupported permutation:[0, 2, 1, 3, 4]: cif
Import error: Unsupported permutation:[0, 2, 1, 3, 4]: 527
Import error: Unsupported permutation:[0, 2, 1, 3, 4]: caf
Import error: Unsupported permutation:[0, 2, 1, 3, 4]: 551
Import error: Unsupported permutation:[0, 2, 1, 3, 4]: 564
Import failed: 6 issues detected
Error while converting model

Screenshot from 2020-09-17 15-31-03

**Comment by svenkreiss** [comment]
@miljanvs Is your engine public?
What kind of permutations are supported?

There is an alternative to the permutations that actually should be faster: in-place ops. They do seem to give me wrong results with onnxruntime, but maybe your engine handles them. Can you try to set `openpifpaf.network.heads.CompositeField3.inplace_ops = False` to True in `export_onnx.py`? That avoids permutations.

**Comment by miljanvs** [comment]
> @miljanvs Is your engine public?

No, it is not a public engine. @mdenna-nviso might have mentioned that we are trying to split the model. We will run base on TesnorRT and heads on our proprietary engine. Now, before splitting the ONNX model, we are trying to run it entirely with our  engine, in order to make sure we have a reference implementation and before making things more complicated with splitting.

> What kind of permutations are supported?
I will have to check and come back to you.

We are trying what you suggested, I will let you know how it went. Thanks.

**Comment by svenkreiss** [comment]
@miljanvs fyi, there is a parallel thread about TensorRT which seems to work (?) with the dev branch: #188

**Comment by miljanvs** [comment]
@svenkreiss , just a brief update... We have managed to get our engine aligned with the 0.12 model version used at the dev branch. The ONNX export test using ONNX Runtime was really helpful. On our engine side, we had to touch a couple of layers but no major changes. Having a solid working reference, we are now looking at model splitting and the TensorRt implementation. Thanks.

**Comment by svenkreiss** [comment]
Hi @miljanvs , that's great to hear. Let me know when I can be of any help.
